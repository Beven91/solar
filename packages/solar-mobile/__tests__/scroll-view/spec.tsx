/**
 * @name ScrollView 测试
 * @description
 */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
// import BScroll from '@better-scroll/core';
import ScrollViewApp from './index.dev';
import ScrollView from '../../src/scroll-view';

const BScroll = require('@better-scroll/core');
const createScroll = BScroll.default;

const ScrollMock = {
  height: 0,
  scrollHeight: 0,
  restore() {
    ScrollMock.height = 0;
    ScrollMock.scrollHeight = 0;
  },
  mockScrollView(height: number, scrollHeight: number) {
    ScrollMock.height = height;
    ScrollMock.scrollHeight = scrollHeight + height;
  },
};
BScroll.default = function(wrapper: HTMLDivElement, options: any) {
  Object.defineProperty(wrapper, 'clientHeight', { value: ScrollMock.height });
  Object.defineProperty(wrapper, 'offsetHeight', { value: ScrollMock.height });
  Object.defineProperty(wrapper.children[0], 'scrollHeight', { value: ScrollMock.scrollHeight });
  Object.defineProperty(wrapper.children[0], 'offsetHeight', { value: ScrollMock.scrollHeight });
  return new createScroll(wrapper, options);
};

describe('ScrollView', () => {
  beforeEach(function() {
    ScrollMock.restore();
    const userAgent = 'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1';
    Object.defineProperty(navigator, 'userAgent', { configurable: true, value: userAgent });
  });

  it('renders correctly', () => {
    const tree = mount(<ScrollViewApp />);
    expect(toJson(tree, { mode: 'deep', noKey: true })).toMatchSnapshot();
  });

  it('scroll', () => {
    const onEnd = jest.fn();
    const onScroll = jest.fn();
    // 模拟滚动条高度
    ScrollMock.mockScrollView(20, 100);
    // 创建渲染树
    const tree = mount(
      <ScrollView
        style={{ height: 20 }}
        onScroll={onScroll}
        onEndReached={onEnd}
      >
        <div style={{ height: 100 }}></div>
      </ScrollView>
    );
    const instance = tree.instance() as ScrollView;
    // 滚动10px
    instance.scrollTo(null, 10);
    // 断言： 在滚动时，需要触发onScroll事件
    expect(onScroll).toBeCalled();
    // 断言：此时还没有后滚动到底，不会触发onEnd
    expect(onEnd).toHaveBeenCalledTimes(0);

    // 滚动到底部
    instance.scrollTo(null, 100);
    // 断言： 在滚动时，需要触发onScroll事件
    expect(onScroll).toBeCalled();
    // 断言：在滚动到底部时，需要触发onEndReached事件
    expect(onEnd).toHaveBeenCalledTimes(1);
  });

  it('destroy', () => {
    // 创建渲染树
    const tree = mount(
      <ScrollView
        style={{ height: 20 }}
      >
        <div style={{ height: 200 }}></div>
      </ScrollView>
    );
    // 销毁
    const instance = tree.instance() as ScrollView;
    tree.unmount();
    expect(instance.scroller).toBeNull();
  });


  it('iscroll.scroller.native.dom', () => {
    // 模拟滚动条高度
    ScrollMock.mockScrollView(20, 100);
    // 创建渲染树
    const tree = mount(
      <ScrollView
        style={{ height: 50 }}
      >
        <div style={{ height: 200 }}></div>
      </ScrollView>
    );

    const instance = tree.instance() as ScrollView;
    const scroller = instance.scroller;
    // 断言：保证存在scroller属性
    expect(scroller).not.toBeNull();

    // 断言：存在scrollHeight 且为一个数值
    expect(scroller.scrollHeight).not.toBeNaN();
    expect(scroller.scrollWidth).not.toBeNaN();
    // 断言：存在scrollTop
    expect(scroller.scrollTop).not.toBeNaN();

    // 添加滚动事件s
    const onScroll = jest.fn();
    scroller.addEventListener('scroll', onScroll);
    // 模拟滚动
    instance.scrollTo(10, 20);
    // 断言：存在addEventListener事件
    expect(onScroll).toHaveBeenCalled();
    // 断言scrollTop值为20
    expect(scroller.scrollTop).toBe(20);
    // 无水平滚动条，所以为0
    expect(scroller.scrollLeft).toBe(0);

    onScroll.mockReset();
    // 移除事件
    scroller.removeEventListener('scroll', onScroll);
    scroller.removeEventListener('scroll', null);
    // 模拟滚动
    instance.scrollTo(null, 30);
    // 断言:onScroll不再被调用，已被移除
    expect(onScroll).not.toHaveBeenCalled();
  });

  it('scrollTo', () => {
    // 模拟滚动条高度
    ScrollMock.mockScrollView(20, 100);
    // 创建渲染树
    const tree = mount(
      <ScrollView
        style={{ height: 50 }}
      >
        <div style={{ height: 200 }}></div>
      </ScrollView>
    );
    const instance = tree.instance() as ScrollView;
    const scroller = instance.scroller;
    // 滚动到指定位置
    instance.scrollTo(10, 20);

    // 断言: 当前设置的值应该生效
    expect(scroller.scrollTop).toBe(20);
    // 无水平滚动条，所以为0
    expect(scroller.scrollLeft).toBe(0);
  });

  it('iScroll.pc', () => {
    const userAgent = 'Mozilla/5.0 (11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Safari/604.1';
    Object.defineProperty(navigator, 'userAgent', { configurable: true, value: userAgent });
    // 创建渲染树
    const tree = mount(
      <ScrollView >
        <div ></div>
      </ScrollView>
    );
    tree.setProps({ onScroll: (a: any) => a });
    const scrollView = tree.instance() as ScrollView;
    const context = {
      scrollTop: 0,
      scrollLeft: 0,
    };
    Object.defineProperty(scrollView.containerRef.current, 'scrollTop', {
      set(v) {
        context.scrollTop = v;
      },
    });
    Object.defineProperty(scrollView.containerRef.current, 'scrollLeft', {
      set(v) {
        context.scrollLeft = v;
      },
    });
    scrollView.scrollTo(1, 20);
    // 断言
    expect(context.scrollLeft).toBe(1);
    expect(context.scrollTop).toBe(20);

    // 定义scrollTo 函数
    Object.defineProperty(scrollView.containerRef.current, 'scrollTo', {
      value: jest.fn(),
    });
    const scrollTo = scrollView.containerRef.current.scrollTo;
    scrollView.scrollTo(1, 20);
    // 断言:
    expect(scrollTo).toHaveBeenCalled();

    tree.unmount();
  });
});