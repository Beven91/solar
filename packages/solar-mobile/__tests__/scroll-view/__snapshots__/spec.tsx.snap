// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollView renders correctly 1`] = `
<div
  className="scroll-view iscroll-viewable vertical "
  onScroll={null}
>
  <div
    className="scroll-inner-view"
  >
    <div
      className="content"
    >
      <p>
         三余无梦生 
        <input
          type="checkbox"
        />
         
      </p>
      <img
        src=""
      />
      <h4>
        表单
      </h4>
      <div
        className="am-list"
      >
        <div
          className="am-list-header"
        >
          CheckboxItem demo
        </div>
        <div
          className="am-list-body"
        >
          <div
            className="am-list-item am-checkbox-item am-list-item-middle"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
          >
            <div
              className="am-list-thumb"
            >
              <label
                className="am-checkbox-wrapper"
              >
                <span
                  className="am-checkbox"
                  style={Object {}}
                >
                  <input
                    checked={false}
                    className="am-checkbox-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                  />
                  <span
                    className="am-checkbox-inner"
                  />
                </span>
              </label>
            </div>
            <div
              className="am-list-line"
            >
              <div
                className="am-list-content"
              >
                Ph.D.
              </div>
            </div>
            <div
              className="am-list-ripple"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
          <div
            className="am-list-item am-checkbox-item am-list-item-middle"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
          >
            <div
              className="am-list-thumb"
            >
              <label
                className="am-checkbox-wrapper"
              >
                <span
                  className="am-checkbox"
                  style={Object {}}
                >
                  <input
                    checked={false}
                    className="am-checkbox-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                  />
                  <span
                    className="am-checkbox-inner"
                  />
                </span>
              </label>
            </div>
            <div
              className="am-list-line"
            >
              <div
                className="am-list-content"
              >
                Bachelor
              </div>
            </div>
            <div
              className="am-list-ripple"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
          <div
            className="am-list-item am-checkbox-item am-list-item-middle"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
          >
            <div
              className="am-list-thumb"
            >
              <label
                className="am-checkbox-wrapper"
              >
                <span
                  className="am-checkbox"
                  style={Object {}}
                >
                  <input
                    checked={false}
                    className="am-checkbox-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                  />
                  <span
                    className="am-checkbox-inner"
                  />
                </span>
              </label>
            </div>
            <div
              className="am-list-line"
            >
              <div
                className="am-list-content"
              >
                College diploma
              </div>
            </div>
            <div
              className="am-list-ripple"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
          <div
            className="am-list-item am-checkbox-item am-checkbox-item-disabled am-list-item-middle"
            data-seed="logId"
            defaultChecked={true}
            onClick={[Function]}
          >
            <div
              className="am-list-thumb"
            >
              <label
                className="am-checkbox-wrapper"
              >
                <span
                  className="am-checkbox am-checkbox-checked am-checkbox-disabled"
                  style={Object {}}
                >
                  <input
                    checked={true}
                    className="am-checkbox-input"
                    disabled={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                  />
                  <span
                    className="am-checkbox-inner"
                  />
                </span>
              </label>
            </div>
            <div
              className="am-list-line am-list-line-multiple"
            >
              <div
                className="am-list-content"
              >
                Undergraduate
                <div
                  className="am-list-brief"
                >
                  Auxiliary text
                </div>
              </div>
            </div>
            <div
              className="am-list-ripple"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
          <div>
            <div
              className="am-list-item am-list-item-middle"
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="am-list-line"
              >
                <div
                  className="am-list-content"
                >
                  Datetime
                </div>
                <div
                  className="am-list-extra"
                >
                  请选择
                </div>
                <div
                  aria-hidden="true"
                  className="am-list-arrow am-list-arrow-horizontal"
                />
              </div>
              <div
                className="am-list-ripple"
                style={
                  Object {
                    "display": "none",
                  }
                }
              />
            </div>
          </div>
          <div
            className="am-list-item am-input-item am-list-item-middle"
          >
            <div
              className="am-list-line"
            >
              <div
                className="am-input-label am-input-label-5"
              >
                光标在左
              </div>
              <div
                className="am-input-control"
              >
                <input
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="start from left"
                  readOnly={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="am-list-item am-input-item am-list-item-middle"
          >
            <div
              className="am-list-line"
            >
              <div
                className="am-input-label am-input-label-5"
              >
                光标在右
              </div>
              <div
                className="am-input-control"
              >
                <input
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="start from right"
                  readOnly={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="am-list-item am-input-item am-list-item-middle"
          >
            <div
              className="am-list-line"
            >
              <div
                className="am-input-label am-input-label-5"
              >
                数字键盘
              </div>
              <div
                className="am-input-control"
              >
                <input
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="money format"
                  readOnly={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="am-list-item am-list-item-middle"
            onClick={[Function]}
          >
            <div
              className="am-list-line"
            >
              <div
                className="am-list-content"
              >
                <p>
                   双刀 
                  <input
                    type="radio"
                  />
                   
                </p>
              </div>
            </div>
            <div
              className="am-list-ripple"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
          <div
            className="am-list-item am-list-item-middle"
            onClick={[Function]}
          >
            <div
              className="am-list-line"
            >
              <div
                className="am-list-content"
              >
                开关
              </div>
              <div
                className="am-list-extra"
              >
                <label
                  className="am-switch"
                >
                  <input
                    checked={false}
                    className="am-switch-checkbox"
                    disabled={false}
                    name=""
                    onChange={[Function]}
                    onClick={[Function]}
                    type="checkbox"
                    value="off"
                  />
                  <div
                    className="checkbox"
                    style={Object {}}
                  />
                </label>
              </div>
            </div>
            <div
              className="am-list-ripple"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
          <div
            className="am-list-item am-list-item-middle"
            onClick={[Function]}
          >
            <div
              className="am-list-line"
            >
              <div
                className="am-list-content"
              >
                输入
              </div>
              <div
                className="am-list-extra"
              >
                <input
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                  type="text"
                />
              </div>
            </div>
            <div
              className="am-list-ripple"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
          <div
            className="am-list-item am-list-item-middle"
            onClick={[Function]}
          >
            <div
              className="am-list-line"
            >
              <div
                className="am-list-content"
              >
                备注
              </div>
              <div
                className="am-list-extra"
              >
                <textarea
                  onChange={[Function]}
                  style={
                    Object {
                      "height": 80,
                      "width": "100%",
                    }
                  }
                  value="华华人间世，浩浩海江衣。知年指荏苒，繁花漫荒初。 洒洒落人间，微微同太古。"
                />
              </div>
            </div>
            <div
              className="am-list-ripple"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
        </div>
      </div>
      <a
        aria-disabled={false}
        className="am-button am-button-primary"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
      >
        <span>
          打开窗口
        </span>
      </a>
      <div>
         
      </div>
      <a
        aria-disabled={false}
        className="am-button am-button-primary"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
      >
        <span>
          跳转到200
        </span>
      </a>
    </div>
    <div>
      <h4>
        轮播组件
      </h4>
      <div
        className="slider am-carousel"
        style={
          Object {
            "MozBoxSizing": "border-box",
            "boxSizing": "border-box",
            "display": "block",
            "height": "auto",
            "position": "relative",
            "visibility": "hidden",
            "width": "100%",
          }
        }
      >
        <div
          className="slider-frame"
          onClick={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "MozBoxSizing": "border-box",
              "WebkitTransform": "translate3d(0, 0, 0)",
              "boxSizing": "border-box",
              "display": "block",
              "height": "auto",
              "margin": "0px",
              "msTransform": "translate(0, 0)",
              "overflow": "hidden",
              "padding": 0,
              "position": "relative",
              "transform": "translate3d(0, 0, 0)",
            }
          }
        >
          <ul
            className="slider-list"
            style={
              Object {
                "MozBoxSizing": "border-box",
                "WebkitTransform": "translate3d(0px, 0px, 0)",
                "boxSizing": "border-box",
                "cursor": "inherit",
                "display": "block",
                "height": 0,
                "margin": "0px 0px",
                "msTransform": "translate(0px, 0px)",
                "padding": 0,
                "position": "relative",
                "transform": "translate3d(0px, 0px, 0)",
                "width": 0,
              }
            }
          >
            <li
              className="slider-slide"
              style={
                Object {
                  "MozBoxSizing": "border-box",
                  "boxSizing": "border-box",
                  "display": "inline-block",
                  "height": "auto",
                  "left": 0,
                  "listStyleType": "none",
                  "marginBottom": "auto",
                  "marginLeft": 0,
                  "marginRight": 0,
                  "marginTop": "auto",
                  "position": "absolute",
                  "top": 0,
                  "verticalAlign": "top",
                  "width": 0,
                }
              }
            >
              <a
                style={
                  Object {
                    "display": "inline-block",
                    "height": 238,
                    "width": "100%",
                  }
                }
              >
                <img
                  alt=""
                  onLoad={[Function]}
                  src=""
                  style={
                    Object {
                      "height": "100%",
                      "verticalAlign": "top",
                      "width": "100%",
                    }
                  }
                />
              </a>
            </li>
            <li
              className="slider-slide"
              style={
                Object {
                  "MozBoxSizing": "border-box",
                  "boxSizing": "border-box",
                  "display": "inline-block",
                  "height": "auto",
                  "left": 0,
                  "listStyleType": "none",
                  "marginBottom": "auto",
                  "marginLeft": 0,
                  "marginRight": 0,
                  "marginTop": "auto",
                  "position": "absolute",
                  "top": 0,
                  "verticalAlign": "top",
                  "width": 0,
                }
              }
            >
              <a
                style={
                  Object {
                    "display": "inline-block",
                    "height": 238,
                    "width": "100%",
                  }
                }
              >
                <img
                  alt=""
                  onLoad={[Function]}
                  src=""
                  style={
                    Object {
                      "height": "100%",
                      "verticalAlign": "top",
                      "width": "100%",
                    }
                  }
                />
              </a>
            </li>
            <li
              className="slider-slide"
              style={
                Object {
                  "MozBoxSizing": "border-box",
                  "boxSizing": "border-box",
                  "display": "inline-block",
                  "height": "auto",
                  "left": 0,
                  "listStyleType": "none",
                  "marginBottom": "auto",
                  "marginLeft": 0,
                  "marginRight": 0,
                  "marginTop": "auto",
                  "position": "absolute",
                  "top": 0,
                  "verticalAlign": "top",
                  "width": 0,
                }
              }
            >
              <a
                style={
                  Object {
                    "display": "inline-block",
                    "height": 238,
                    "width": "100%",
                  }
                }
              >
                <img
                  alt=""
                  onLoad={[Function]}
                  src=""
                  style={
                    Object {
                      "height": "100%",
                      "verticalAlign": "top",
                      "width": "100%",
                    }
                  }
                />
              </a>
            </li>
          </ul>
        </div>
        <div
          className="slider-decorator-0"
          style={
            Object {
              "bottom": 0,
              "position": "absolute",
              "textAlign": "center",
              "width": "100%",
            }
          }
        >
          <div
            className="am-carousel-wrap"
          >
            <div
              className="am-carousel-wrap-dot am-carousel-wrap-dot-active"
            >
              <span
                style={Object {}}
              />
            </div>
            <div
              className="am-carousel-wrap-dot"
            >
              <span
                style={Object {}}
              />
            </div>
            <div
              className="am-carousel-wrap-dot"
            >
              <span
                style={Object {}}
              />
            </div>
          </div>
        </div>
        <style
          dangerouslySetInnerHTML={
            Object {
              "__html": ".slider-slide > img {width: 100%; display: block;}",
            }
          }
          type="text/css"
        />
      </div>
    </div>
    <div>
      <h4>
        嵌套滚动
      </h4>
      <div
        className="scroll-view iscroll-viewable vertical "
        onScroll={null}
        style={
          Object {
            "height": 400,
          }
        }
      >
        <div
          className="scroll-inner-view"
        >
          <p>
            素还真
          </p>
          <img
            src=""
          />
          <p>
            天踦爵
          </p>
          <img
            src=""
          />
        </div>
      </div>
    </div>
    <div>
      <h4>
        水平滚动
      </h4>
      <div
        className="scroll-view iscroll-viewable horizon horizon"
        onScroll={null}
      >
        <div
          className="scroll-inner-view"
        >
          <div
            className="container"
          >
            <div
              className="inner"
            />
            <div
              className="inner2"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="content yellow"
    />
    <div
      className="content blue"
    />
    ~到底啦....
  </div>
</div>
`;
